samtranslator/model/eventsources/push.py

```diff
@@ -218,6 +218,8 @@ class S3(PushEventSource):
     def resources_to_link(self, resources):
         if isinstance(self.Bucket, dict) and "Ref" in self.Bucket:
             bucket_id = self.Bucket["Ref"]
+            if not isinstance(bucket_id, str):
+                raise InvalidEventException(self.relative_id, "'Ref' value in S3 events is not a valid string.")
```

Comment:
The error messages seems not match the condition. `bucket_id` here is not a str, while the error message makes me feel like it is a str, just not a valid one. Does the situation here mean the `Ref` key is missing?

Reply:
Ref is a parameter name which should have some value. like Ref: somevalue . Bucket_id is that value. Therefore we are checking bucket_id which should be a string i.e. value of 'Ref' in the template should be a string. Make sense?

Reply:
will `bucket_id` be other types other than `str` and `None`?

Reply:
No. I should only be string.

Reply:
What about Python 2 where a string can be `unicode`?

Reply:
> No. It should only be string.

Reply:
> No. It should only be string.

Can I assume if `not isinstance(bucket_id, str)` is `True`, it means the user forgets to put a `Ref`, and as long as `Ref` exists, the `not isinstance(bucket_id, str)` is `False`?

Reply:
no. The user can have other types like list or dict as well as value for Ref. e.g.
Ref:
  -abc
or
Ref:
  abc:123

Reply:
we mainly care for hashable values here.

Reply:
Oh got it, thanks for the explanation!

Reply:
resolving this. python 2 related unicode discussion is here https://github.com/aws/serverless-application-model/pull/1757#discussion_r513794030